### PR TITLE
Sign uninstall.exe

### PR DESCRIPTION
**Pull request by SinghRajenM**
There were numerous threads/issues which reported that npp's `uninstall.exe` is detected by various antivirus.

However, those were false positives. Mainly, these were detected by ML module which works upon static characteristics of binary. Sometimes, these ML modules detect just `Hello World` program too. Unfortunately, there are some limitation to handle these cases. But signing characteristics of binary play a very vital role there (due to signature based scanning/rules). Hence, valid signed binaries can help to mitigate these false positives.

So, it is good if we can sign the uninstaller exe. Prior to `nsis 3.08`, it was a complex process to sign the uninstaller binary. But `nsis version 3.08` makes it very easy with the help of `uninstfinalize`.  Refer [nsis documentation](https://nsis.sourceforge.io/Docs/Chapter5.html#uninstfinalize).

Additionally, it can sign installer too using `finalize`. Just an improvement to automating things.

Related issues #4120, #5806, #4443. May be few more.

@donho  further things to take care - 
1. Replace `npp_sign_cert.pfx` with actual/test npp certificate. I know original cert file can not be in repo. Hence, 
    1. You can generate test cert and put it in repo.
    2. Or you can define a macro like `SINGING_NEEDED` and both the `finalize` and `uninstfinalize` can be guarded using it.
    3. Either of these two will help others as nsis will not complain at compilation.
5. You can grab latest SignTool.exe from Widows SDK kit (e.g. `C:\Program Files (x86)\Windows Kits\10\bin\10.0.22621.0\x86\signtool.exe`). This may require to update sign script too as there could be some new parameter.

### CODE REVIEW

File: PowerEditor/installer/sign-binaries.bat [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+".\signCert\SignTool.exe" sign /f ".\signCert\npp_sign_cert.pfx" "%1"
+if errorlevel 1 goto SigningFailed
+goto SigningOK
+
+:SigningFailed
+echo Failed to sign file %1
+exit 1
+
+:SigningOK
```
Comment (donho) [inline PowerEditor/installer/sign-binaries.bat:13]: Please rename the file `sign-binaries.bat` to `sign-installers.bat`,
and use the following content instead:
 
```batch
@ECHO OFF

if [%SIGN%] == [] goto NoSignInstaller
if not %SIGN% == 1 goto NoSignInstaller

ECHO Start signing file: %1
%signBinary% "%1"

if errorlevel 1 goto SigningFailed
goto SigningOK

:SigningFailed
echo Failed to sign file %1
exit 1

:NoSignInstaller

:SigningOK
```

`%signBinary%` is already defined in `packageAll.bat`

File: PowerEditor/installer/signCert/npp_sign_cert.pfx [hunk 2]
```diff
@@ -0,0 +1 @@
+Replace with your actual cert or test cert
```
Comment (donho) [inline PowerEditor/installer/signCert/npp_sign_cert.pfx:1]: Remove both `npp_sign_cert.pfx` and `SignTool.exe` files.
They are not necessary, and should not be in the repository.

File: PowerEditor/installer/nppSetup.nsi [hunk 3]
```diff
@@ -43,6 +43,10 @@ OutFile ".\build\npp.${APPVERSION}.Installer.arm64.exe"
 OutFile ".\build\npp.${APPVERSION}.Installer.exe"
 !endif
 
+; Sign both installer and uninstaller
+!finalize        'sign-binaries.bat "%1"' = 0     ; %1 is replaced by the installer exe to be signed.
+!uninstfinalize  'sign-binaries.bat "%1"' = 0     ; %1 is replaced by the uninstaller exe to be signed.
```
Comment (donho) [inline PowerEditor/installer/nppSetup.nsi:48]: 1. Rename the file sign-binaries.bat to sign-installers.bat.
2. Remove the following lines from [packageAll.bat](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/installer/packageAll.bat#L545) :
``` batch
if %SIGN% == 0 goto NoSignInstaller
%signBinary% !nppInstallerVar!
If ErrorLevel 1 goto End
%signBinary% !nppInstallerVar64!
If ErrorLevel 1 goto End
%signArmBinary% !nppInstallerVarArm64!
If ErrorLevel 1 goto End
:NoSignInstaller
```

**Review by donho — CHANGES_REQUESTED** [review]
Please consider the following suggestions:

File: PowerEditor/installer/sign-binaries.bat (hunk 1 above)
Comment (SinghRajenM) [reply PowerEditor/installer/sign-binaries.bat]: It is not that easy to get the `%signBinary%` from `packageAll.bat` -> `nsis compiler` -> then `sign-installers.bat`.

However, I have made some changes which should work (I could not test as I don't have cert and password handy). Also, I assume that you must have two env variable for `%NPP_CERT%` and `%NPP_CERT_PWD%`.

File: PowerEditor/installer/nppSetup.nsi [hunk 4]
```diff
@@ -43,6 +43,12 @@ OutFile ".\build\npp.${APPVERSION}.Installer.arm64.exe"
 OutFile ".\build\npp.${APPVERSION}.Installer.exe"
 !endif
 
+; Sign both installer and uninstaller
+!ifdef SIGN_INSTALLER
```
Comment (SinghRajenM) [inline PowerEditor/installer/nppSetup.nsi:47]: This compiler flag is passed from `PackageAll.bat`. That means if installer is created by just compiling `nppSetup.nsi` (e.g. right click on file and compile with nsis), then both installer and uninstaller will not be signed.

File: PowerEditor/installer/packageAll.bat [hunk 5]
```diff
@@ -477,10 +477,23 @@ If ErrorLevel 1 goto End
...
+  set nsisPath="%PROGRAMFILES%\NSIS\makensis.exe"
+)
+
+set sign_flag=-DSIGN_INSTALLER
+if [%SIGN%]==[] (
+  set "sign_flag="
+)
+
+rem build x86, x64 and arm64 installer
+%nsisPath% /V4 %sign_flag% nppSetup.nsi > nsis_compile_log_x86.txt
```
Comment (SinghRajenM) [inline PowerEditor/installer/packageAll.bat:494]: Dump nsis compilation log to file. Easy for debugging. 😊

**Comment by SinghRajenM** [comment]
> Please use the suggestions after my first review for the following reasons:
> 
> 1. They are tested on my prod machine.
> 2. They are simpler & less code.

I wonder how you tested and how it was working. When I was testing , neither %SIGN% nor %signBinary% were coming to `sing_installer.bat`. I echoed them, both were empty. Moreover, I was not getting what is happening when nsis compiler was running. Hence, /V4 flag & log redirection to file.


``` batch
@ECHO OFF

if [%SIGN%] == [] goto NoSignInstaller
if not %SIGN% == 1 goto NoSignInstaller

ECHO Start signing file: %1
%signBinary% "%1"

if errorlevel 1 goto SigningFailed
goto SigningOK

:SigningFailed
echo Failed to sign file %1
exit 1

:NoSignInstaller

:SigningOK
```

**Comment by donho** [comment]
> I wonder how you tested and how it was working. When I was testing , neither %SIGN% nor %signBinary% were coming to sing_installer.bat. I echoed them, both were empty. Moreover, I was not getting what is happening when nsis compiler was running. Hence, /V4 flag & log redirection to file.

I fill %SIGN% and other variables (password & certif file) under cmd.exe before launching packageAll.exe.
There's no way to pass these varibles to nsis GUI compiler, so when users use nsis GUI compiler, the installers are not signed.
If any user want to sign Notepad++ installer, they have to use packageAll.bat way.

**Comment by SinghRajenM** [comment]
> > I wonder how you tested and how it was working. When I was testing , neither %SIGN% nor %signBinary% were coming to sing_installer.bat. I echoed them, both were empty. Moreover, I was not getting what is happening when nsis compiler was running. Hence, /V4 flag & log redirection to file.
> 
> I fill %SIGN% and other variables (password & certif file) under cmd.exe before launching packageAll.bat (`set SIGN=1`). There's no way to pass these varibles to nsis GUI compiler, so when users use nsis GUI compiler, the installers are not signed. If any user want to sign Notepad++ installer, they have to use packageAll.bat way.

Done!

**Comment by donho** [comment]
Thank you @SinghRajenM !

**Pull request closed without merging**